### PR TITLE
feat(contry-selector): add favorites countries & few others props

### PR DESCRIPTION
**Pull request by emri99**
* add properties :
  * `favoriteCountries` : list of isoCode to be displayed on top of the list
  * `sortCountries` : whether countries should be sorted by localised name
  * `addFavoritesSeparator` : whether a divider should be added between favorites and defaults countries
  * `showCountryDialCode` : whether the dial code should be displayed as ListTile subtitle
  * `noResultMessage` : the message displayed instead of the list when the search has no results
* add proxy properties for above properties in selector navigator classes
* add API docs on updated files
* fix countryFinder mismatch order on cleaning search after performing a search

Remaining things to do :
* [ ] write tests
* [ ] write translations for `noResultMessage`
* [ ] write documentation
* [ ] update example to handle favorites countries

### CODE REVIEW

**Comment by cedvdb** [comment]
Merged on dev_favorites branch as I'll help a bit on this one for some of the remaining tasks

File: lib/src/widgets/country_picker/country_selector.dart [hunk 1]
```diff
@@ -1,22 +1,60 @@
...
+  /// list divider between favorites and others defaults countries
+  final List<String> favoritesCountries;
+
+  /// Whether to add a list divider between favorites & defaults
+  /// countries.
+  final bool addSeparator;
+
+  /// Whether to show the country dial code (ie: +1 / +33 /...)
+  /// as a listTile subtitle
+  final bool showDialCode;
```
Comment (cedvdb) [inline lib/src/widgets/country_picker/country_selector.dart:40]: It has been brought to my attention that "dial code" is a bit ambiguous and "country dial code" might be preferable. I'm letting a comment here for future reference.

File: lib/src/widgets/country_picker/country_selector.dart [hunk 2]
```diff
@@ -1,22 +1,60 @@
...
 
+  /// Sort the countries automatically by localized name.
+  /// Note that the favorites countries are not sorted but
+  /// displayed in defined order.
+  final bool sortCountries;
+
+  /// Determine the countries to be displayed on top of the list
+  /// Check [addSeparator] property to enable/disable adding a
+  /// list divider between favorites and others defaults countries
+  final List<String> favoritesCountries;
```
Comment (cedvdb) [inline lib/src/widgets/country_picker/country_selector.dart:32]: should be `favoriteCountries`

File: lib/src/widgets/country_picker/country_selector.dart [hunk 3]
```diff
@@ -1,22 +1,60 @@
...
+
+  /// Whether to add a list divider between favorites & defaults
+  /// countries.
+  final bool addSeparator;
```
Comment (cedvdb) [inline lib/src/widgets/country_picker/country_selector.dart:36]: This is ambiguous. I'd rename to something that clearly states that the divider is to be placed between favorites and defaults.

File: lib/src/widgets/country_picker/country_selector.dart [hunk 4]
```diff
@@ -25,17 +63,62 @@ class CountrySelector extends StatefulWidget {
 
 class _CountrySelectorState extends State<CountrySelector> {
   late List<Country> _filteredCountries;
-  late final _countryFinder;
+  late CountryFinder _countryFinder;
+  int? _separatorIndex;
```
Comment (cedvdb) [inline lib/src/widgets/country_picker/country_selector.dart:67]: this is ambiguous as per the comment above, I'd rename it to _favoritesSeparatorIndex. It's ambiguous in the case we add another separator in the future for example a separator between each countries (with no favorites).

File: lib/src/widgets/country_picker/country_selector.dart (hunk 2 above)
Comment (emri99) [reply lib/src/widgets/country_picker/country_selector.dart:32]: updated ;)

File: lib/src/widgets/country_picker/country_selector.dart (hunk 1 above)
Comment (emri99) [reply lib/src/widgets/country_picker/country_selector.dart:40]: updated

File: lib/src/widgets/country_picker/country_selector.dart (hunk 3 above)
Comment (emri99) [reply lib/src/widgets/country_picker/country_selector.dart:36]: I don't find an appropriate property. My current idea is quite a long term  `addSeparatorAfterFavorites`... 

Any better idea would be greatly appreciated ;)

File: lib/src/widgets/country_picker/country_selector.dart (hunk 4 above)
Comment (emri99) [reply lib/src/widgets/country_picker/country_selector.dart]: updated

File: lib/src/widgets/country_picker/country_selector.dart (hunk 3 above)
Comment (cedvdb) [reply lib/src/widgets/country_picker/country_selector.dart:36]: addFavoritesSeparator should do it

Comment (emri99) [reply lib/src/widgets/country_picker/country_selector.dart]: updated ;)

**Comment by emri99** [comment]
I'm currently writing tests, but I've got a weird issue.
I have 2 tests successful when running them standalone, but they are failing in the whole test suite.
When failing, the error is due to the `ListTile` widgets not found at all. I've try to add a delay of 1s or 10s using `tester.pump` method.
Do you have already experiment this ? any ideas of what may cause this ?
Thanks ;)

**Comment by cedvdb** [comment]
I did actually, I'm not even sure it was for this package in particular, it could be though. I did not find a solution except by running them in standalone mode.

**Comment by emri99** [comment]
I doubt this is an issue with flutter itself as we would probably not be the only ones in this case and I cannot find anything in this sense on the Internet.
Will investigate more on what should cause this in tested widget ;)

**Comment by emri99** [comment]
For currently unknown reason, the `PhoneFieldLocalization.delegate` is in cause.

Here's the builder I use in each test : 

```dart
final builder = ({
        List<String>? favorites,
        bool addFavoritesSeparator = false,
      }) =>
          MaterialApp(
            locale: Locale('fr', ''),
            localizationsDelegates: [
              ...GlobalMaterialLocalizations.delegates,
              PhoneFieldLocalization.delegate,
            ],
            supportedLocales: [Locale('fr', '')],
            home: Scaffold(
              body: CountrySelector(
                onCountrySelected: (c) {},
                addFavoritesSeparator: addFavoritesSeparator,
                favoriteCountries: favorites ?? <String>[],
                sortCountries: true,
              ),
            ),
          );
```

If I comment the `PhoneFieldLocalization.delegate,`, no errors.

**Comment by cedvdb** [comment]
I will investigate tomorrow to find the root cause. If you want you can submit with the test separated.

**Comment by emri99** [comment]
I've push problematic test on a new branch of my fork (`test-issue`), I hope this can be useful for you to investigate the problem ;) test is partially commented, so that above issue is the only reason for test to fail ;)

**Comment by cedvdb** [comment]
For what it's worth it finds the MaterialApp but not the Scaffold. I've also found a similar issue in google discussion but no answers and it dates.

**Comment by cedvdb** [comment]
This also fails when more than one test is run from this group:

```dart
      final builder = ({
        List<String>? favorites,
        bool addFavoritesSeparator = false,
      }) =>
          MaterialApp(
            locale: Locale('en'),
            localizationsDelegates: [
              ...GlobalMaterialLocalizations.delegates,
              PhoneFieldLocalization.delegate,
            ],
            supportedLocales: [Locale('en')],
            home: Scaffold(
              body: Container(),
            ),
          );

      testWidgets('should be properly sorted without favorites',
          (tester) async {
        await tester.pumpWidget(builder());
        await tester.pumpAndSettle();
        final scaffold = find.byType(Scaffold);
        expect(scaffold, findsOneWidget);
      });
      ```

**Comment by emri99** [comment]
You're first message states that scaffold is not found, was it a mistake ?
Do you try to comment usage of `PhoneFieldLocalization.delegate,` ?

**Comment by cedvdb** [comment]
In the snippet I posted above, Scaffold fails to be found (there is nothing in the app but a scaffold). Therefor I believe this is an issue with flutter itself

**Comment by cedvdb** [comment]
However I did write a huge map just to see if it passes with precompiled arb files and it does pass. Therefor I believe flutter unit testing is not loading json files properly or it's interfering with the tree build.

I'm not sure where to go from there. To convert to arb or not.

**Comment by emri99** [comment]
Honestly, I'm not sure to understand neither what/why you've tried nor the consequences 🤫 🤷‍♂️

This may make sense to push a specific branch here as a reproductible case for creating an issue on flutter itself.

**Comment by cedvdb** [comment]
I'm going to merge this one in a specific branch for posting a flutter issue

**Comment by cedvdb** [comment]
Leaking testWidgets issue:

https://github.com/flutter/flutter/issues/90079

**Comment by cedvdb** [comment]
@emri99 I've added your test to the branch localization_arb_fix they are now passing. I let the comment that were there. So you can now see the results of tests.

For information the loading process for translation is making the test fail. The official documentation does it another way, by generating dart classes, which makes test pass as there is nothing asynchronous. However! there is an option in their generator 
to also use deferred file loading like I was doing and this option also makes test fail. So this is definitely a flutter unit test framework problem.

**Pull request merged**